src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -1460,3 +1496,91 @@ void Terminal::_updateUrlDetection()
         ClearPatternTree();
     }
 }
+
+void Terminal::AddMark(const Microsoft::Console::VirtualTerminal::DispatchTypes::ScrollMark& mark,
+                       const til::point& start,
+                       const til::point& end)
+{
+    DispatchTypes::ScrollMark m = mark;
+    m.start = start;
+    m.end = end;
+
+    _scrollMarks.push_back(m);

```

Comment:
I see now... the list is not sorted.
We could make it sorted, so that we can write less code elsewhere and use more efficient algorithms. Not many containers have efficient middle-insertions, apart from stuff like binary trees, but binary trees are basically linked lists, so... storing potentially thousands of marks in the auto-mark case might perform very poorly. We could use a [rope](https://en.wikipedia.org/wiki/Rope_(data_structure)) like structure which would work like a `std::deque` but were we'd preserve some "padding" between chunks so that we can insert items without having to shift the entire list. But we'd have to write one first...
If we want to use sorted marks right now a `std::set` might work best.

A way to make this scalable and work well with the auto-mark mode would be by making the feature worse: Instead of storing marks as "ranges" of rows so that they can be cleared atomically as a whole, we could make scroll marks "row-wise": When you mark row 15 to 20 you create 6 marks and not 1. This would allow us to - conceptually speaking - store such marks as attributes on our TextBuffer's `ROW`s. Sorted insertion, invalidation during buffer circling, etc. aren't a problem anymore, because we fundamentally circumvented these problems.

Solving the performance of updating the scrollbar with marks can also be solved this way: We can "downsample" our 9000 row marks by - let's say - 10x and create another `std::vector<uint8_t>` of size 900. Each item in the smaller vector then counts the cumulative amount of marks in it's bucket of 10 rows. In a sense this would work exactly like a [LOD/MIPMAP](https://en.wikipedia.org/wiki/Mipmap). That way we can keep scrollbar updates performant...

I'd be in favor of such an approach as it would likely reduce the amount of code and scale well in the future.

Reply:
Lots to unpack here:

>  make scroll marks "row-wise": When you mark row 15 to 20 you create 6 marks and not 1

I specifically wanted marks to not just be an entire row. Granted, in this PR they mostly are. But I wanted folks to be able to select some text and "bookmark" it for later. Or combine this with regex matching and let people automatically mark `[eE]rror:.*` blobs as Error marks. Or even the FTCS sequences - I envisioned the whole set of A-D sequences all correlate to one single mark on the prompt (so that the D sequence could change that one mark to an Error/Success mark).

I had a bit of discussion earlier that this is like, a parallel feature with the hyperlinks right now. They're both "a range of text that indicates _something_". I think there's room in the future to consolidate these in a way that also makes room for more arbitrary regex matches. Of course, this is still experimental, which is why I didn't hash that out yet. 

notes for later: Hyperlinks are stored as a `uint16_t id` in `TextAttribute`. Supposedly marks, regexes could go in that `id` as well, we'd just need the `_hyperlinkMap` in `TextBuffer` to point to values that are more than just a string URI. And then we'd use that value to determine how to render it. Sticking marks into there would get us resize/reflow for marks "for free". That has knock-onn effects though on how text gets retrieved from the buffer - hyperlinks need to be styled as "different" attributes, but marks/regexes don't _necessarily_ need to be.

Reply:
Oh right... reflow. That's another thing. I'm fine with the current solution. 🙂

Hyperlinks btw also deserve an improved implementation in my opinion: I'm having this problem were if I view my diffs without pager, it causes the entire terminal to stutter during scrolling, because we use an O(n*m) algorithm to prune hyperlinks on newlines:
https://github.com/microsoft/terminal/blob/11b810e4036b669e5f223cce447f76d56044be0f/src/buffer/out/textBuffer.cpp#L1383-L1398

(`GetHyperlinks` itself is also `O(n)` on top of that.)

Also I haven't quite understood so far how we're storing hyperlinks if we got more than 64k. I guess it just doesn't work? In either case, making bookmarks or marks in general scale well, will certainly require some serious (algorithm) investment, because hyperlinks right now aren't working well either unfortunately. 😓